src/codec/meta/vector_clock/clock_inner.rs

```diff
@@ -0,0 +1,197 @@
+/// filesystem and is used for providing strict ordering of events. The internal value is
+/// initialized to a random value when a new one is initialized.
+///
+/// Internally this uses an atomic 64 unsigned integer for the ID, wrapping is an allowed behavior
+/// and must be handled by all consumers. We consider a roll over valid if the last known ID was
+/// within 262,144 (2^18) ticks of rolling over.
+
+#[derive(Debug, Clone)]
+pub struct ClockInner(Arc<AtomicU64>);
+
+impl ClockInner {
+    pub fn initialize() -> Self {
+        // TODO: make this actually initialize to a random value as the docs above indicate
+        Self::from(ClockInnerSnapshot::from(0))
+    }
+
+    pub fn to_snapshot(&self) -> ClockInnerSnapshot {
+        ClockInnerSnapshot::from(self)
+    }
+}
+
+impl From<ClockInnerSnapshot> for ClockInner {
+    fn from(val: ClockInnerSnapshot) -> Self {
+        Self(Arc::new(AtomicU64::new(val.0)))
+    }
+}
+
+impl PartialEq for ClockInner {
+    fn eq(&self, other: &Self) -> bool {
+        self.0
+            .load(Ordering::Relaxed)
+            .eq(&other.0.load(Ordering::Relaxed))
+    }
+}
+
+const WRAP_THRESHOLD: u64 = 2 ^ 18;
+
+/// A snapshot of a [`VectorClock`] at a specific value
+///
+/// These are what get stored to record the state of a vector
+/// during specific operations
+///
+/// # Wrapping Behavior
+/// These must functionally monotonically increase, but if we overflow
+/// the underlying value we will wrap around. This is handled by
+/// comparing the values with a threshold to determine if the
+/// wrapped value is greater than the non-wrapped value.
+/// The threshold is 2^18, or 262,144.
+/// [`PartialOrd`] and [`Ord`] are implemented to handle this.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct ClockInnerSnapshot(pub(super) u64);
+
+impl ClockInnerSnapshot {
+    pub async fn encode<W: AsyncWrite + Unpin + Send>(
+        &self,
+        writer: &mut W,
+    ) -> std::io::Result<usize> {
+        let clock_bytes = self.0.to_le_bytes();
+
+        writer.write_all(&clock_bytes).await?;
+
+        Ok(clock_bytes.len())
+    }
+
+    pub fn parse(input: Stream) -> ParserResult<Self> {
+        let (input, value) = le_u64.parse_peek(input)?;
+        Ok((input, Self(value)))
+    }
+
+    pub const fn size() -> usize {
+        8
+    }
+}
+
+impl From<&ClockInner> for ClockInnerSnapshot {
+    fn from(value: &ClockInner) -> Self {
+        Self(value.0.load(Ordering::Relaxed))
+    }
+}
+
+impl From<u64> for ClockInnerSnapshot {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+
+impl PartialOrd for ClockInnerSnapshot {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ClockInnerSnapshot {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        if self.0 < WRAP_THRESHOLD || other.0 < WRAP_THRESHOLD {
```

Comment:
I suspect we need to have two threshold checks here comparing against `u64::MAX - WRAP_THRESHOLD`. I see you added tests around these so I could be wrong. Definitely just a "gut" feeling.

Reply:
I'm not 100% sure what what two threshold checks you mean. I'm like 98% sure that implementation is correct, but if you can think of any failure cases then I will be able to handle them